frontend/packages/pipelines-plugin/src/components/shared/common/auto-complete/useAutoComplete.ts

```diff
@@ -76,96 +90,189 @@ const useFilterOptions = (options: string[]): FilterOptionsHook => {
+    },
+    [options],
+  );
 
   return [filteredOptions, setFilter];
 };
 
 /**
- * Manages auto complete state.
+ * Listens to the node in various ways to prefer functions of AutoComplete.
  */
-const useAutoComplete = (
-  autoCompleteValues: string[],
-  onAutoComplete: (newValue: string) => void,
-): AutoCompleteHook => {
+const useNodeListener = (
+  cursorPosition: React.MutableRefObject<CursorPosition>,
+  menuOptions: MenuKeyWatcherHook,
+  filterOptions: FilterOptionsHook,
+  setFocusingOptions: (isFocusing: boolean) => void,
+  closeCleanup: () => void,
+): [React.MutableRefObject<ElementType>, (node: ElementType) => void] => {
   const [node, setNode] = React.useState<ElementType>(null);
   const nodeRef = React.useRef<ElementType>(null);
   nodeRef.current = node;
-  const cursorPosition = React.useRef<CursorPosition>([0, 0]);
-  const { isOpen, shouldOpen, close } = useOpenMenuKeyWatcher();
-  const [options, setFilter] = useFilterOptions(autoCompleteValues);
+
+  const { isOpen, shouldOpen } = menuOptions;
+  const [, setFilter] = filterOptions;
 
   const onKeyCallback = React.useCallback(
     (e) => {
+      const applyFilterAtCursor = () => {
+        const filterValue = nodeRef.current.value.substring(
+          cursorPosition.current[0],
+          cursorPosition.current[1],
+        );
+        setFilter(filterValue);
+      };
+
       if (isOpen) {
         cursorPosition.current = [cursorPosition.current[0], nodeRef.current.selectionEnd];
-        const filterValue = nodeRef.current.value.substr(...cursorPosition.current);
-        setFilter(filterValue);
+
+        if (cursorPosition.current[1] < cursorPosition.current[0]) {
+          // User moved before the start of the autoComplete trigger, close
+          closeCleanup();
+          return;
+        }
+
+        applyFilterAtCursor();
       } else {
         cursorPosition.current = [nodeRef.current.selectionStart, nodeRef.current.selectionEnd];
-        shouldOpen(e, (additionalContent = '') => {
-          onAutoComplete(
-            insertIntoValue(nodeRef.current.value, cursorPosition.current, additionalContent),
-          );
-          // Align the start to be after any additional characters so we can track just new content
-          const startPos = cursorPosition.current[0];
-          const newStartPos = startPos + additionalContent.length;
-          cursorPosition.current = [newStartPos, newStartPos];
-        });
+        const isOpening = shouldOpen(e);
+
+        if (isOpening) {
+          // In the case they are highlighting text, invoke auto complete immediately on the text
+          applyFilterAtCursor();
+        }
       }
     },
-    [isOpen, setFilter, shouldOpen, onAutoComplete],
+    [isOpen, setFilter, shouldOpen, closeCleanup, cursorPosition],
+  );
+
+  const focusDropdownCallback = React.useCallback(
+    (e) => {
+      if (!isOpen) return;
+
+      if (e.code === 'Tab' || e.code === 'ArrowDown') {
```

Comment:
Pressing tab focuses the first item in the dropdown, but it is not cycling through items in the list. is this intended?

Reply:
Damn, nope that is not intended. I thought it did tbh... must have been an earlier implementation. Will have to look into that. Thanks!

Reply:
Looking into this, it's actually a PF component that prevents the tab... not sure this would be easy to fix given the available API to it.